### ISSUE TITLE
Code stops working if I comment away debug() 

### ISSUE THREAD

**Issue by Snorvarg** [cakephp/cakephp#12942]
This is a bug.

* CakePHP Version: 3.5.17
* Platform and Target: Using Windows 10 v1803, IIS v10.0.17134.1, MySql v5.7.23-log.

Tables: (just the relevant columns included)
```
categories
  id
  parent_id
```

```
cat_lang
  id
  category_id
  i18n
  url_title
```

Models:
`CategoriesTable (having hasMany('CatLang'))`
and 
`CatLangTable (having belongsTo('Categories'))`

### What you did

This is the query. I try to fetch the unique dataset where Categories.parent_id, CatLang.i18n and CatLang.url_title matches the given variables. 

```
    $query = $this->find('all', 
      [
        'fields' => ['Categories.id', 'Categories.parent_id']
      ])
      ->matching('CatLang', function(Query $q) use($i18n, $urlTitle){
        $q = $q->where(['CatLang.i18n' => $i18n, 'CatLang.url_title' => $urlTitle]);

        // This one!!
        debug($q);
        
        return $q;
      })
      ->where(['parent_id is' => null]);
    
    $category = $query->all();
```

### What happened

If I turn off the debug($q); where I have commented "This one!!" I get an empty result. The $category will contain an empty row with nulled variables. 

But if I leave the debug() on, it works. Then $category contains the expected object, as below:

```
items' => [
		(int) 0 => object(Cake\ORM\Entity) {
			'id' => (int) 14,
			'parent_id' => null,
			'_matchingData' => [
				'CatLang' => object(Cake\ORM\Entity) {

					'id' => (int) 17,
					'category_id' => (int) 14,
					'i18n' => 'sv',
					'url_title' => 'home',
                          ......
```

This is the row I get if debug() is turned off:

```
object(Cake\ORM\ResultSet) {
	'items' => [
		(int) 0 => object(Cake\ORM\Entity) {
			'id' => (int) 14,
			'parent_id' => null,
			'[new]' => false,
			'[accessible]' => [
				'*' => true
			],
			'[dirty]' => [],
			'[original]' => [],
			'[virtual]' => [],
			'[errors]' => [],
			'[invalid]' => [],
			'[repository]' => 'Categories'
		}
	]
}
```

**Comment by ADmad** [comment]
Debugging a query object will execute the query, so results won't be same as executing the query after complete query is built.

The output you have shown without the debug is expected and you are only selecting `id` and `parent_id` fields and hence you won't have any other fields in the generated resultset entities.

**Comment by Snorvarg** [comment]
Ok, that makes sense. I know this is not a help forum, and I have solved this with a join(), but how am I supposed to make my example above work with matching()? I mean, get the _matchingData-array as in my example when the debug() is around.
I have followed docs pretty close, the examples with matching() specifically tell me to write the query this way. 

I assume the solution is not to leave debug() hanging where it is in my example... :-)
